src/main/java/teammates/common/util/TzdbResourceZoneRulesProvider.java

```diff
@@ -0,0 +1,181 @@
+package teammates.common.util;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.StreamCorruptedException;
+import java.lang.reflect.Method;
+import java.time.zone.ZoneOffsetTransition;
+import java.time.zone.ZoneOffsetTransitionRule;
+import java.time.zone.ZoneRules;
+import java.time.zone.ZoneRulesException;
+import java.time.zone.ZoneRulesProvider;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Loads time-zone rules for 'TZDB'. Modified from {@link java.time.zone.TzdbZoneRulesProvider}.
+ * Reads the file 'tzdb.dat' from resources instead of the JRE library directory.
+ */
+public final class TzdbResourceZoneRulesProvider extends ZoneRulesProvider {
+
+    /** Type for ZoneRules. */
+    private static final byte ZRULES = 1;
+    /** Type for ZoneOffsetTransition. */
+    private static final byte ZOT = 2;
+    /** Type for ZoneOffsetTransition. */
+    private static final byte ZOTRULE = 3;
+
+    /**
+     * All the regions that are available.
+     */
+    private List<String> regionIds;
+    /**
+     * Version Id of this tzdb rules.
+     */
+    private String versionId;
+    /**
+     * Region to rules mapping.
+     */
+    private final Map<String, Object> regionToRules = new ConcurrentHashMap<>();
```

Comment:
Is there a more specific class to be used here rather than Object?

Reply:
Oops maybe I didn't make it clear enough; everything added in the first commit was copied from the JDK 8 source code for `TzdbZoneRulesProvider`. Hence reviews for this class should just be about whether the behaviour looks exactly the same after making the adaptations in the second commit.

Reply:
Granted, but this is an area that can be improved without affecting the correctness of the code, right? Although this is just a minor stuff that can be ignored.

Reply:
It's difficult to change as they put a mix of `byte[]`, `ZoneRules`, `ZoneOffsetTransition` and `ZoneOffsetTransitionRule` into this map. I won't attempt to decipher the logic here.

Reply:
In that case `Object` seems the safest way out.